OpenRA.Mods.Common/Traits/Player/TimeLimitManager.cs

```diff
@@ -0,0 +1,133 @@
+
+		[Desc("Time Limit options that will be shown in the lobby dropdown. Values are in minutes.")]
+		public readonly int[] TimeLimitOptions = { 0, 20, 30, 40, 60, 90 };
+
+		[Desc("List of remaining minutes of game time when a text and optional speech notification should be made to players.")]
+		public readonly Dictionary<int, string> TimeLimitWarnings = new Dictionary<int, string>
+		{
+			{ 1, null },
+			{ 2, null },
+			{ 3, null },
+			{ 4, null },
+			{ 5, null },
+			{ 10, null },
+		};
+
+		[Desc("Default selection for the time limit option in the lobby. Should use one of the TimeLimitOptions.")]
+		public readonly int TimeLimitDefault = 0;
+
+		[Desc("Prevent the time limit option from being changed in the lobby.")]
+		public readonly bool TimeLimitLocked = false;
+
+		[Desc("Display order for the time limit dropdown in the lobby.")]
+		public readonly int TimeLimitDisplayOrder = 0;
+
+		IEnumerable<LobbyOption> ILobbyOptions.LobbyOptions(Ruleset rules)
+		{
+			var timelimits = TimeLimitOptions.ToDictionary(c => c.ToString(), c =>
+			{
+				if (c == 0)
+					return "No limit";
+				else
+					return c.ToString() + " minute{0}".F(c > 1 ? "s" : null);
+			});
+
+			yield return new LobbyOption("timelimit", TimeLimitLabel, TimeLimitDescription, true, TimeLimitDisplayOrder,
+				new ReadOnlyDictionary<string, string>(timelimits), TimeLimitDefault.ToString(), TimeLimitLocked);
+		}
+
+		public object Create(ActorInitializer init) { return new TimeLimitManager(init.Self, this); }
+	}
+
+	public class TimeLimitManager : INotifyTimeLimit, ITick
+	{
+		TimeLimitManagerInfo info;
+		public readonly int TimeLimit;
+
+		public TimeLimitManager(Actor self, TimeLimitManagerInfo info)
+		{
+			this.info = info;
+
+			var tl = self.World.LobbyInfo.GlobalSettings.OptionOrDefault("timelimit", info.TimeLimitDefault.ToString());
+			if (!int.TryParse(tl, out TimeLimit))
+				TimeLimit = info.TimeLimitDefault;
+
+			// Convert from minutes to ticks
+			TimeLimit *= 60 * (1000 / self.World.Timestep);
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (TimeLimit <= 0)
+				return;
+
+			var ticksPerSecond = 1000 / self.World.Timestep;
+			var ticksRemaining = TimeLimit - self.World.WorldTick;
+
+			if (ticksRemaining > 0 && info.TimeLimitWarnings.Keys.All(m => ticksRemaining != m * 60 * ticksPerSecond))
+				return;
+
+			var ntls = self.TraitsImplementing<INotifyTimeLimit>();
+
+			if (ticksRemaining == 0)
+			{
+				foreach (var ntl in ntls)
+					ntl.NotifyTimerExpired(self);
+
+				return;
+			}
+
+			foreach (var ntl in ntls)
+				foreach (var t in info.TimeLimitWarnings.Keys)
+					if (ticksRemaining == t * 60 * ticksPerSecond)
+						ntl.NotifyTimeRemaining(self, t, info.TimeLimitWarnings[t]);
+		}
+
+		void INotifyTimeLimit.NotifyTimeRemaining(Actor self, int minutesRemaining, string audioNotification)
+		{
+			if (self.Owner == self.World.RenderPlayer || self.World.RenderPlayer == null)
```

Comment:
This causes N (= # of players) messages to be written into the chat for observers.
The audio notification doesn't seem to work in RA either, but its not clear why.